YakimS/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,231 @@
+package CodeUAss1Package;
+
+import javax.management.openmbean.KeyAlreadyExistsException;
+import java.util.*;
+
+/**
+ * This class contains the root to a binary tree that can have only
+ * one node for each possible key.
+ * @param <T> is the type of data stored in the tree nodes value
+ */
+public class BinaryTree<T> {
+
+    private Node root;
+    private Set<Integer> keySet;
+
+    public BinaryTree() {
+        keySet = new HashSet<>();
+    }
+
+    public void printBinaryTree()
+    {
+        printBinaryTree(this.root,0);
+    }
+    private void printBinaryTree(Node root, int level){
+        if(root==null)
+            return;
+        printBinaryTree(root.getRightSon(), level+1);
+        if(level!=0){
+            for(int i=0;i<level-1;i++)
+                System.out.print("|\t\t");
+            System.out.println("|-------"+root.getKey());
+        }
+        else
+            System.out.println(root.getKey());
+        printBinaryTree(root.getLeftSon(), level+1);
+    }
+
+    /**
+     * This method add a node to a random valid place in the binary tree
+     * @param value - value of the added node
+     * @param key - key of the added node
+     * @throws KeyAlreadyExistsException if there is already a node with key as given in the parameter
+     */
+    public void add(T value, int key) throws KeyAlreadyExistsException
+    {
+        if(keySet.contains(key))
+        {
+            throw new KeyAlreadyExistsException();
+        }
+        else {
+            keySet.add(key);
+        }
+
+        Node nodeToAdd = new Node(key, value);
+        if(root == null)
+        {
+            root = nodeToAdd;
+            return;
+        }
+
+        /*
+        *   Moving down the tree, while choosing the path down (through right or left son)
+        *  randomly, till the current node is null, and then assign to the added node
+        *  to that null current node
+        */
+        Node currentNode = root;
+        Random random = new Random();

```

Comment:
Using a randomized data structure is an interesting idea, but it makes it very hard to test and to reason about. Unit tests are not required for this assignment, but if they were, it would be nearly impossible to check if they are correct, and since you use `Random` without seed, the results you will get are very likely to be different each time you run the program.